llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp

```diff
@@ -0,0 +1,574 @@
+  assert(MI->getOpcode() == EVM::FCALL && "Unexpected call instruction");
+  assert(CurMBB == MI->getParent());
+
+  // PUSH_LABEL technically increases the stack height on 1, but we don't
+  // increase it explicitly here, as the label will be consumed by the following
+  // JUMP.
+  StackHeight += StackAdj;
+
+  // Create pseudo jump to the callee, that will be expanded into PUSH and JUMP
+  // instructions in the AsmPrinter.
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoCALL))
+                   .addGlobalAddress(Func);
+
+  // If this function returns, we need to create a label after JUMP instruction
+  // that is followed by JUMPDEST and this is taken care in the AsmPrinter.
+  // In case we use setPostInstrSymbol here, the label will be created
+  // after the JUMPDEST instruction, which is not what we want.
+  if (WillReturn)
+    NewMI.addSym(CallReturnSyms.at(MI));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitRet(const MachineInstr *MI) {
+  assert(MI->getOpcode() == EVM::RET && "Unexpected ret instruction");
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoRET));
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitUncondJump(
+    const MachineInstr *MI, MachineBasicBlock *Target) {
+  assert(MI->getOpcode() == EVM::JUMP &&
+         "Unexpected unconditional jump instruction");
+  assert(StackHeight >= 0);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoJUMP))
+                   .addMBB(Target);
+  verify(NewMI);
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitCondJump(
+    const MachineInstr *MI, MachineBasicBlock *Target) {
+  assert(MI->getOpcode() == EVM::JUMPI &&
+         "Unexpected conditional jump instruction");
+  StackHeight -= 1;
+  assert(StackHeight >= 0);
+  auto NewMI = BuildMI(*CurMBB, CurMBB->end(), MI->getDebugLoc(),
+                       TII->get(EVM::PseudoJUMPI))
+                   .addMBB(Target);
+  verify(NewMI);
+}
+
+// Verify that a stackified instruction doesn't have registers and dump it.
+void EVMStackifyCodeEmitter::CodeEmitter::verify(const MachineInstr *MI) const {
+  assert(EVMInstrInfo::isStack(MI) &&
+         "Only stackified instructions are allowed");
+  assert(all_of(MI->operands(),
+                [](const MachineOperand &MO) { return !MO.isReg(); }) &&
+         "Registers are not allowed in stackified instructions");
+
+  LLVM_DEBUG(dbgs() << "Adding: " << *MI << "stack height: " << StackHeight
+                    << "\n");
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::finalize() {
+  for (MachineBasicBlock &MBB : MF)
+    for (MachineInstr &MI : make_early_inc_range(MBB))
+      // Remove all the instructions that are not stackified.
+      // FIXME: Fix debug info for stackified instructions and don't
+      // remove debug instructions.
```

Comment:
Let's create a ticket for this if it doesn't exist.

Reply:
Created a ticked and added ticket number.